daemon/cmd/deletion_queue.go

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cmd
+
+import (
+	"context"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/cilium/cilium/pkg/defaults"
+	"github.com/cilium/cilium/pkg/lock/lockfile"
+)
+
+// processQueuedDeletes is the agent-side of the identity deletion queue.
+// The CNI plugin queues deletions when the agent is down, because
+// containerd / crio expect CNI DEL to always succeed. It does so
+// by writing a file to /run/cilium/deleteQueue with the endpoint ID
+// to delete.
+//
+// On startup, we will grab a lockfile in this directory, then process
+// all deletions. Then, we start up the agent server, then drop the lock.
+// Any CNI processes waiting in that period of time will, after getting
+// the lock.
+//
+// Returns a done function that drops the lock. It should be called
+// after the server is running.
+func (d *Daemon) processQueuedDeletes() func() {
+	if err := os.MkdirAll(defaults.DeleteQueueDir, 0755); err != nil {
+		log.WithError(err).WithField("path", defaults.DeleteQueueDir).Error("Failed to ensure deletion queue directory exists")
+		return func() {}
+	}
+
+	log.Infof("Processing queued endpoint deletions")
+
+	var lf *lockfile.Lockfile
+	locked := false
+
+	unlock := func() {
+		if lf != nil && locked {
+			lf.Unlock()
+		}
+		if lf != nil {
+			lf.Close()
+		}
+	}
+
+	lf, err := lockfile.NewLockfile(defaults.DeleteQueueLockfile)
+	if err != nil {
+		log.WithError(err).WithField("filename", defaults.DeleteQueueLockfile).Error("Failed to create ")
+		return unlock
+	}
+
+	ctx, cancel := context.WithTimeout(d.ctx, 10*time.Second)
+	defer cancel()
+	err = lf.Lock(ctx, true)
+	if err != nil {
+		log.WithError(err).Error("Failed to lock deletion directory, proceeding anyways")
```

Comment:
```suggestion
		log.WithError(err).Warn("Failed to lock CNI deletion directory, proceeding anyways")
```

What are the consequences of this failure? What should the user know?

Reply:
I've been thinking about this. On the one hand, it means we might miss an endpoint deletion in the worst possible case. On the other hand, I really **really** don't want to ever block agent startup for something non-critical. Plus, we might as well proceed with cleanup regardless.

One possible error message would be "Failed to lock CNI deletion directory, proceeding anyways. This may cause CNI deletions to be missed" which sounds drastic but is technically correct.

Reply:
I think it's fair enough.